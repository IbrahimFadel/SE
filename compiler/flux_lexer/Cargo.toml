[package]
name = "flux_lexer"
version = "0.1.0"
edition = "2021"

[dependencies]
logos = "0.14.0"
text-size = "1.1.1"
salsa = { path = "../../lib/salsa" }
