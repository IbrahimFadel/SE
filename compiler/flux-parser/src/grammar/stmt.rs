use flux_lexer::T;

use super::{expr::type_expr, *};

pub(crate) fn stmt(p: &mut Parser) -> CompletedMarker {
	if p.at(T!(let))
		|| p.at(T!(iN))
		|| p.at(T!(uN))
		|| p.at(T!(f32))
		|| p.at(T!(f64))
		|| (p.at(T!(ident)) && matches!(p.peek_next(), Some(T!(ident)) | Some(T!(eq))))
		|| p.at(T!(lparen))
	{
		var_decl(p)
	} else if p.at(T!(return)) {
		return_stmt(p)
	} else {
		let m = p.start();
		expr::expr(p, true);
		if p.at(T!(semicolon)) {
			p.bump();
		}
		m.complete(p, SyntaxKind::ExprStmt)
	}
}

fn return_stmt(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	p.bump();
	if p.at(T!(semicolon)) {
		p.bump();
		return m.complete(p, SyntaxKind::ReturnStmt);
	}
	expr::expr(p, true);
	p.expect(
		T!(semicolon),
		format!("expected `;` after return statement"),
	);
	m.complete(p, SyntaxKind::ReturnStmt)
}

// fn else_if_stmt(p: &mut Parser) -> CompletedMarker {
// 	let m = p.start();
// 	p.expect(T!(if), format!("expected `if` in else if statement"));
// 	expr::expr(p, false);
// 	block(p);
// 	m.complete(p, SyntaxKind::IfStmt)
// }

fn var_decl(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	if p.at(T!(let)) {
		p.bump();
	} else {
		type_expr(p);
	}
	p.expect(
		TokenKind::Ident,
		format!("expected identifier in variable declaration"),
	);
	p.expect(
		TokenKind::Eq,
		format!("expected `=` in variable declaration"),
	);
	expr::expr(p, true);
	p.expect(
		TokenKind::SemiColon,
		format!("expected `;` after variable declaration"),
	);
	m.complete(p, SyntaxKind::VarDecl)
}

#[cfg(test)]
mod tests {
	use crate::test_stmt_str;
	test_stmt_str!(var_decl, "i32 x = 1;");
	test_stmt_str!(
		var_decl_fcont,
		r#"i32 y =
	i32 x = 1;"#
	);
}
