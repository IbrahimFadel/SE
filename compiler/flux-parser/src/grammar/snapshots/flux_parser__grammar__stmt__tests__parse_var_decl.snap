---
source: compiler/flux-parser/src/grammar/stmt.rs
expression: parse.to_string()
---
VarDecl@0..10
  LetKw@0..3 "let"
  Whitespace@3..4 " "
  Ident@4..5 "x"
  Whitespace@5..6 " "
  Eq@6..7 "="
  Whitespace@7..8 " "
  IntExpr@8..9
    IntExpr@8..9 "1"
  SemiColon@9..10 ";"
